### PR TITLE
2022 02 08 ROS2 launch HAL

### PR DESCRIPTION
**Pull request by zultron**
This PR is based on #13 and should be reviewed after that PR's resolution.

Launching ROS nodes in HAL presents a few challenges:
- HAL RT components can't be directly started with the `Node` launch action
  - RT comps are DLL plugins instead of command-line executables, as the `Node` action expects
  - RT comps are traditionally loaded from a HAL config file, with no direct way to pass ROS command-line arguments
- HAL configuration bringup requires certain ordering:
  - The HAL real-time environment must be started first
  - Before a HAL component's pins may be linked to signals, it must first be loaded and waited on to become ready

This PR introduces a series of new launch actions for HAL intended to solve these problems:
- Abstract launch action classes to help ordering HAL configuration actions:
  - `HalOrderedAction`:  The basic action for ordering execution:  This action defers execution until receiving a `HalReady` event from the previous action, and after completing execution emits another `HalReady` event to trigger the next action
  - `HalAsyncReadyAction`:  For actions that take significant time to execute, this action avoids blocking the launch process by periodically checking for execution complete, emitting a `HalReady` event or, upon exceeding a timeout, a `Shutdown` event
  - `HalThreadedReadyAction`:  Similar to above, actions that take significant time to execute, but in a Python routine, avoid blocking similarly, but running the Python routine in a separate thread

- `HalConfig`:  Based on the `GroupAction`, this action groups the `HalOrderedAction`s that make up a complete HAL configuration:
  - `HalMgr`:  Based on the `Node` action, the `HalConfig` automatically starts this first; the `HalMgr` action run the `hal_mgr` node, responsible for starting and stopping the HAL realtime environment
  - `HalRTNode`:  Loads a RT HAL component and waits for it to call `hal_ready()` before emitting the `HalReady` event; based on the `Node` launch action, ROS params are passed using the same familiar launch arguments
  - `HalUserNode`:  Loads a (non-RT) user component and waits for it to become ready; based on `Node` launch action & uses familiar launch arguments
  - `HalFiles`:  Loads one or more Python or `.hal` HAL files, which may load additional RT or user components, create signals and net component pins, create and start RT threads with component functions added
 
This PR also updates `hal_mgr` and other scripts to use the new system, as well as the `hal_rrbot_control` launchers, removing the old, awkward and delicate plumbing.

### CODE REVIEW

**Comment by zultron** [comment]
Parent PR #13 is merged, thank you @vatanaksoytezer !
This PR is rebased & tests are passing.  Ready for review.

File: hal_hw_interface/hal_hw_interface/launch/hal_ordered_action.py [hunk 1]
```diff
@@ -0,0 +1,315 @@
...
+
+    @property
+    def hal_name(self):
+        return self.__hal_name
+
+    def cancel(self, context, warn=True):
+        """
+        Cancel this Action's handlers.
+
+        This function is not thread-safe and should be called only from
```
Comment (vatanaksoytezer) [inline hal_hw_interface/hal_hw_interface/launch/hal_ordered_action.py]: Is there anything that prevents us from making this thread safe?

File: hal_rrbot_control/launch/rrbot.launch.py [hunk 2]
```diff
@@ -185,117 +176,112 @@ def generate_launch_description():
...
+                component="hal_control_node",
+                parameters=[
+                    dict(robot_description=robot_description_content),
+                    LaunchConfiguration("robot_controller_config"),
+                ],
+                log_cmd=True,
+                output="both",
+                emulate_tty=True,
+                # arguments=['--ros-args', '--log-level', 'debug'],
+                # prefix=['xterm -e gdb -ex run --args'],
```
Comment (vatanaksoytezer) [inline hal_rrbot_control/launch/rrbot.launch.py:249]: Maybe delete some unuused comments here

**Review by vatanaksoytezer — COMMENTED** [review]
Left a few comments that I couldn't understand. Let me know if these makes sense. Rest looks good to me

File: hal_hw_interface/hal_hw_interface/launch/hal_ordered_action.py (hunk 1 above)
Comment (zultron) [reply hal_hw_interface/hal_hw_interface/launch/hal_ordered_action.py]: This function started life as copy'n'paste from the `ros2/launch` `timer_action`:

https://github.com/ros2/launch/blob/cac43b195fb767128df1fdf494687973d538c792/launch/launch/actions/timer_action.py#L113

The current incarnation has nothing in common with that code except this docstring!  I'll remove it.

BUT, answering your question, very little of the internals of the ROS2 launch system is thread safe, and most everything runs from an ioloop coroutine out of the main thread.  Everything in this PR follows the usual conventions of the ROS2 launch internals, and as such, all the code runs out of the main ioloop.  The only exception to this is in the `HalThreadedReadyAction` class, which launches actions that must be executed as blocking Python calls; in this case, the Python call is run in a separate thread, and the thread synchronization is all in thread-safe HAL functions.

File: hal_rrbot_control/launch/rrbot.launch.py (hunk 2 above)
Comment (zultron) [reply hal_rrbot_control/launch/rrbot.launch.py:249]: These are actually left intentionally.  The `hal_rrbot_control` package serves mainly as a demo for folks wanting to use this code.  These options can be uncommented to enable debug logs or to start `gdb`.

**Comment by zultron** [comment]
Thanks for the review, @vatanaksoytezer !  I added a commit to clean up after your first comment.  If you're happy with my explanations & fix, could you resolve the conversations & approve the PR?

**Review by vatanaksoytezer — APPROVED** [review]
Sounds good to me, I'm happy to get this merged as is.

**Pull request merged**